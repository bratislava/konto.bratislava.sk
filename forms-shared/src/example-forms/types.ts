import { GenericObjectType } from '@rjsf/utils'
import { FormsBackendFile } from '../form-files/serverFilesTypes'
import { ClientFileInfo } from '../form-files/fileStatus'

export type ExampleForm<FormData = GenericObjectType> = {
  name: string
  formData: FormData
  serverFiles?: FormsBackendFile[]
  clientFiles?: ClientFileInfo[]
  sharepointFieldMap?: Record<string, string>
}
