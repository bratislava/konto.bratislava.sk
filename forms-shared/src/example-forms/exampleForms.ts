import priznanieKDaniZNehnutelnostiExample1 from './examples/priznanieKDaniZNehnutelnostiExample1'
import priznanieKDaniZNehnutelnostiExample5 from './examples/priznanieKDaniZNehnutelnostiExample5'
import priznanieKDaniZNehnutelnostiExample2 from './examples/priznanieKDaniZNehnutelnostiExample2'
import priznanieKDaniZNehnutelnostiExample3 from './examples/priznanieKDaniZNehnutelnostiExample3'
import priznanieKDaniZNehnutelnostiExample4 from './examples/priznanieKDaniZNehnutelnostiExample4'
import predzahradkyExample from './examples/predzahradkyExample'
import { ExampleForm } from './types'
import zavazneStanoviskoKInvesticnejCinnostiExample from './examples/zavazneStanoviskoKInvesticnejCinnostiExample'
import komunitneZahradyExample from './examples/komunitneZahradyExample'
import stanoviskoKInvesticnemuZameruExample from './examples/stanoviskoKInvesticnemuZameruExample'

export const exampleForms: Record<string, ExampleForm[]> = {
  'stanovisko-k-investicnemu-zameru': [stanoviskoKInvesticnemuZameruExample],
  'zavazne-stanovisko-k-investicnej-cinnosti': [zavazneStanoviskoKInvesticnejCinnostiExample],
  predzahradky: [predzahradkyExample],
  'komunitne-zahrady': [komunitneZahradyExample],
  'priznanie-k-dani-z-nehnutelnosti': [
    priznanieKDaniZNehnutelnostiExample1,
    priznanieKDaniZNehnutelnostiExample2,
    priznanieKDaniZNehnutelnostiExample3,
    priznanieKDaniZNehnutelnostiExample4,
    priznanieKDaniZNehnutelnostiExample5,
  ],
}

export const exampleDevForms: Record<string, ExampleForm[]> = {
  'ziadost-o-pridelenie-najomneho-bytu': [],
}
