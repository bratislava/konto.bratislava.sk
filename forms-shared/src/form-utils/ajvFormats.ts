import { validate as validateUuid, version as uuidVersion } from 'uuid'
import { electronicFormatIBAN, validateIBAN } from 'ibantools'
// https://stackoverflow.com/a/51177696
export const baTimeRegex = /^(\d|0\d|1\d|2[0-3]):[0-5]\d$/

/**
 * Compares two time strings in format HH:MM.
 *
 * Copied from: https://github.com/ajv-validator/ajv-formats/blob/4dd65447575b35d0187c6b125383366969e6267e/src/formats.ts#L180C1-L186C2
 */
function compareBaTime(s1: string, s2: string): number | undefined {
  if (!(s1 && s2)) return undefined
  const t1 = new Date(`2020-01-01T${s1}`).valueOf()
  const t2 = new Date(`2020-01-01T${s2}`).valueOf()
  if (!(t1 && t2)) return undefined
  return t1 - t2
}

// https://blog.kevinchisholm.com/javascript/javascript-e164-phone-number-validation/
export const baPhoneNumberRegex = /^\+[1-9]\d{10,14}$/

// https://github.com/mnestorov/regex-patterns?tab=readme-ov-file#slovakia
// But `libphonenumber` contains stricter validation rules:
// https://github.com/google/libphonenumber/blob/e3dbc4bde172a49d06e86b8305a7ca624e95cae8/resources/PhoneNumberMetadata.xml#L27018
export const baSlovakPhoneNumberRegex = /^\+421[1-9][0-9]{8}$/

export const parseRatio = (value: string) => {
  const ratioRegex = /^(0|[1-9]\d*)\/([1-9]\d*)$/
  if (!ratioRegex.test(value)) {
    return { isValid: false }
  }

  const parts = value.split('/')
  const numerator = parseInt(parts[0], 10)
  const denominator = parseInt(parts[1], 10)

  if (numerator > denominator) {
    return { isValid: false }
  }

  return { isValid: true, numerator, denominator }
}

export const validateBaFileUuid = (value: unknown): value is string => {
  return typeof value === 'string' && validateUuid(value) && uuidVersion(value) === 4
}

const baAjvInputFormats = {
  'ba-slovak-zip': /\b\d{5}\b/,
  'ba-phone-number': baPhoneNumberRegex,
  'ba-slovak-phone-number': baSlovakPhoneNumberRegex,
  'ba-ratio': {
    validate: (value: string) => parseRatio(value).isValid,
  },
  'ba-ico': /^\d{6,8}$/,
  'ba-iban': {
    validate: (value: string) => {
      const electronicFormat = electronicFormatIBAN(value)
      if (!electronicFormat) {
        return false
      }
      return validateIBAN(electronicFormat).valid
    },
  },
}

export type BaAjvInputFormat = keyof typeof baAjvInputFormats

const baAjvOtherFormats = {
  'ba-time': {
    // https://stackoverflow.com/a/51177696
    validate: baTimeRegex,
    compare: compareBaTime,
  },
  'ba-file-uuid': {
    validate: validateBaFileUuid,
  },
}

export const baAjvFormats = {
  ...baAjvInputFormats,
  ...baAjvOtherFormats,
}
