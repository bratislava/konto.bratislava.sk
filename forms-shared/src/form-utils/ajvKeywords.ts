import 'json-schema';
import { JSONSchema7 } from 'json-schema'
declare module 'json-schema' {
  export interface JSONSchema7 {
    pospID?: string
    pospVersion?: string
    slug?: string
    hash?: string
    stepperTitle?: string
    file?: boolean
  }
}

export type BAJSONSchema7 = JSONSchema7

export const baAjvKeywords = [
  // Root schema
  {
    keyword: 'pospID',
  },
  {
    keyword: 'pospVersion',
  },
  {
    keyword: 'slug',
  },
  // Step schema
  {
    keyword: 'hash',
  },
  {
    keyword: 'stepperTitle',
  },
  // File field schema
  {
    keyword: 'file',
  },
]
