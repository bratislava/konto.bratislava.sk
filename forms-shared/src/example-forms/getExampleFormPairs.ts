import { FormDefinition } from '../definitions/formDefinitionTypes'
import {
  getFormDefinitionBySlug,
  getFormDefinitionBySlugDev,
} from '../definitions/getFormDefinitionBySlug'
import flatten from 'lodash/flatten'
import { exampleDevForms, exampleForms } from './exampleForms'
import { ExampleForm } from './types'

type ExampleFormPair<FD = FormDefinition> = {
  formDefinition: FD
  exampleForm: ExampleForm
}

export function getExampleFormPairs<FD extends FormDefinition>({
  formDefinitionFilterFn = (formDefinition: FormDefinition): formDefinition is FD => true,
  includeDevForms = false,
}: {
  formDefinitionFilterFn?: (formDefinition: FormDefinition) => formDefinition is FD
  includeDevForms?: boolean
} = {}): ExampleFormPair<FD>[] {
  const getFormDefinitionBySlugLocal = includeDevForms
    ? getFormDefinitionBySlugDev
    : getFormDefinitionBySlug
  const exampleFormsLocal = includeDevForms ? { ...exampleForms, ...exampleDevForms } : exampleForms

  const slugs = Object.keys(exampleFormsLocal)
  const formDefinitions = slugs
    .map(getFormDefinitionBySlugLocal)
    .filter(
      (formDefinition) => formDefinition != null && formDefinitionFilterFn(formDefinition),
    ) as FD[]

  const withExampleForms = formDefinitions.map((formDefinition) => {
    const exampleFormsBySlug = exampleFormsLocal[formDefinition.slug]
    return exampleFormsBySlug.map((exampleForm) => ({ formDefinition, exampleForm }))
  })

  return flatten(withExampleForms)
}
