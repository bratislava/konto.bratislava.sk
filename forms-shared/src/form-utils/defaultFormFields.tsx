import { FieldProps } from '@rjsf/utils'
import { getDefaultRegistry } from '@rjsf/core'
import React, { ComponentType, useMemo } from 'react'
import { BAJSONSchema7 } from './ajvKeywords'
import { defaultUiSchema } from './formDefaults'
import { getObjectFieldInfo } from './getObjectFieldInfo'

const assertBaOrderInProperties = (
  entries: [string, BAJSONSchema7][],
): entries is [string, { baOrder: number }][] => {
  return entries.every(([key, value]) => {
    return typeof key === 'string' && typeof value === 'object' && typeof value.baOrder === 'number'
  })
}

const defaultRegistry = getDefaultRegistry()
const { SchemaField, ObjectField } = defaultRegistry.fields

/**
 * Extracts `baUiSchema` from the schema and uses it as `uiSchema`.
 *
 * The RJSF says that "The `SchemaFieldRender` component is the work-horse of react-jsonschema-form", this is the top-level
 * component that renders each subschema in the form. Therefore, it is possible to extract the UI schema from the schema
 * here and let the original `SchemaField` act as it would be originally there.
 */
const ExtractUiSchemaSchemaField = (props: FieldProps) => {
  const { isFormObject } = getObjectFieldInfo(props.idSchema)
  const baUiSchema = props.schema.baUiSchema
  const uiSchemaEmptyObject = props.uiSchema ? Object.keys(props.uiSchema).length === 0 : false

  // `uiSchema` should be present only in the root object (whether `defaultUiSchema` or empty)
  if (isFormObject && (props.uiSchema === defaultUiSchema || uiSchemaEmptyObject)) {
    return <SchemaField {...props} />
  }

  if (props.uiSchema) {
    throw new Error('uiSchema is not supported')
  }

  return <SchemaField {...props} uiSchema={baUiSchema} />
}

/**
 * Dynamically creates `ui:order` based on `baOrder` values in schema.
 *
 * The input schema of this function is resolved object schema with merged all of subschemas (see fastMergeAllOf).
 */
const OrderObjectField = (props: FieldProps) => {
  if (props.uiSchema?.['ui:order']) {
    throw new Error('ui:order should not be present in the original uiSchema')
  }

  const newUiSchema = useMemo(() => {
    const entries = Object.entries(props.schema.properties as Record<string, BAJSONSchema7>)
    if (!assertBaOrderInProperties(entries)) {
      throw new Error('baOrder should be present in all properties')
    }

    const entriesSorted = entries.toSorted(
      ([, child1], [, child2]) => child1.baOrder - child2.baOrder,
    )
    const uiOrder = entriesSorted.map(([key]) => key)

    return {
      ...props.uiSchema,
      'ui:order': uiOrder,
    }
  }, [props.uiSchema, props.schema.properties])

  return <ObjectField {...props} uiSchema={newUiSchema} />
}

export type DefaultFormFieldType = 'SchemaField' | 'ObjectField'

export const defaultFormFields = {
  SchemaField: ExtractUiSchemaSchemaField,
  ObjectField: OrderObjectField,
} satisfies Record<DefaultFormFieldType, ComponentType<FieldProps>>
