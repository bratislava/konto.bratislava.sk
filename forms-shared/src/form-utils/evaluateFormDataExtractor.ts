import { GenericObjectType } from '@rjsf/utils'
import { BAJSONSchema7 } from './ajvKeywords'
import { getBaRjsfValidator } from './validators'

type ExtractFn<FormDataType extends GenericObjectType> = (formData: FormDataType) => string

/**
 * Data extractors with schema should be called before the data are validated with form schema
 * (e.g. when displaying form subject in list of forms). The `schema` is validated and upon
 * success the form data are extracted. The extraction function relies on the schema validation,
 * the extraction should never fail or return `undefined` (in that case the function throws an error).
 */
export type SchemaFormDataExtractor<FormDataType extends GenericObjectType> = {
  type: 'schema'
  extractFn: ExtractFn<FormDataType>
  schema: BAJSONSchema7
  schemaValidationFailedFallback?: string
}

/**
 * Data extractors without schema should be called after the data are validated with form schema
 * (e.g. when assigning subject to Ginis, extracting email from form data). The extraction function
 * relies on the fact the data are already validated, the extraction should never fail or return
 * `undefined` (in that case the function throws an error).
 */
export type SchemalessFormDataExtractor<FormDataType extends GenericObjectType> = {
  type: 'schemaless'
  extractFn: ExtractFn<FormDataType>
}

export function evaluateFormDataExtractor<FormDataType extends GenericObjectType>(
  extractor: SchemalessFormDataExtractor<FormDataType>,
  formData: GenericObjectType,
): string
export function evaluateFormDataExtractor<FormDataType extends GenericObjectType>(
  extractor: SchemaFormDataExtractor<FormDataType>,
  formData: GenericObjectType,
  schemaValidationFailedParentFallback: string,
): string
export function evaluateFormDataExtractor<FormDataType extends GenericObjectType>(
  extractor: SchemalessFormDataExtractor<FormDataType> | SchemaFormDataExtractor<FormDataType>,
  formData: GenericObjectType,
  schemaValidationFailedParentFallback?: string,
): string {
  if (extractor.type === 'schema') {
    const { schema, extractFn, schemaValidationFailedFallback } = extractor
    const validator = getBaRjsfValidator()

    if (validator.isValid(schema, formData, schema)) {
      const result = extractFn(formData as FormDataType)
      if (result == null) {
        throw new Error(
          `Extraction returned an empty string for form data: ${JSON.stringify(formData)}.`,
        )
      }

      return result
    }

    return schemaValidationFailedFallback ?? schemaValidationFailedParentFallback!
  } else if (extractor.type === 'schemaless') {
    const { extractFn } = extractor
    const result = extractFn(formData as FormDataType)
    if (result == null) {
      throw new Error(
        `Extraction returned an empty string for form data: ${JSON.stringify(formData)}.`,
      )
    }

    return result
  } else {
    throw new Error('Invalid extractor type')
  }
}
