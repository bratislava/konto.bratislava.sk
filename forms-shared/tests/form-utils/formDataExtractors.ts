import { describe, expect, test } from 'vitest'
import { getExampleFormPairs } from '../../src/example-forms/getExampleFormPairs'
import {
  extractEmailFormEmail,
  extractEmailFormName,
  extractFormSubjectPlain,
  extractFormSubjectTechnical,
} from '../../src/form-utils/formDataExtractors'
import {
  FormDefinition,
  FormDefinitionEmail,
  FormDefinitionSlovenskoSkGeneric,
  FormDefinitionType,
  isEmailFormDefinition,
  isSlovenskoSkGenericFormDefinition,
} from '../../src/definitions/formDefinitionTypes'
import { SchemalessFormDataExtractor } from '../../src/form-utils/evaluateFormDataExtractor'

describe('formDataExtractors', () => {
  describe('extractFormSubjectPlain', () => {
    test('should return form definition title if subject.extractPlain is not provided', () => {
      const formDefinition = {
        title: 'Test Form Title',
      } as FormDefinition
      const formData = {}

      const result = extractFormSubjectPlain(formDefinition, formData)
      expect(result).toBe('Test Form Title')
    })

    test('should return form title if formData is null', () => {
      const extractSubjectMock: SchemalessFormDataExtractor<any> = {
        type: 'schemaless',
        extractFn: (formData) => 'Extracted Subject',
      }

      const formDefinition = {
        title: 'Test Form Title',
        subject: { extractPlain: extractSubjectMock },
      } as unknown as FormDefinition
      const formData = null

      const result = extractFormSubjectPlain(formDefinition, formData)
      expect(result).toBe('Test Form Title')
    })

    test('should return result of subject.extractPlain if provided and formData is not null', () => {
      type FormDataType = {
        subject: string
      }
      const extractSubjectMock: SchemalessFormDataExtractor<FormDataType> = {
        type: 'schemaless',
        extractFn: (formData) => formData.subject,
      }

      const formDefinition = {
        title: 'Test Form Title',
        subject: {
          extractPlain: extractSubjectMock,
        },
      } as unknown as FormDefinitionSlovenskoSkGeneric
      const formData: FormDataType = { subject: 'Extracted Subject' }

      const result = extractFormSubjectPlain(formDefinition, formData)
      expect(result).toBe('Extracted Subject')
    })

    test('should throw error when extractFormSubjectPlain returns empty result', () => {
      type FormDataType = {
        subject: string
      }
      const extractSubjectMock: SchemalessFormDataExtractor<FormDataType> = {
        type: 'schemaless',
        extractFn: (formData) => formData.subject,
      }

      const formDefinition = {
        title: 'Test Form Title',
        subject: { extractPlain: extractSubjectMock },
      } as unknown as FormDefinition
      const formData = {
        wrongSubject: 'Extracted Subject',
      }

      expect(() => extractFormSubjectPlain(formDefinition, formData)).toThrow()
    })

    getExampleFormPairs().forEach(({ formDefinition, exampleForm }) => {
      test(`${exampleForm.name} form subject should match snapshot`, () => {
        const result = extractFormSubjectPlain(formDefinition, exampleForm.formData)
        expect(result).toMatchSnapshot()
      })
    })
  })

  describe('extractFormSubjectTechnical', () => {
    test('should return form definition title if extractFormSubjectTechnical is not provided', () => {
      const formDefinition = {
        title: 'Test Form Title',
        ginisAssignment: {},
      } as FormDefinitionSlovenskoSkGeneric
      const formData = {}

      const result = extractFormSubjectTechnical(formDefinition, formData)
      expect(result).toBe('Test Form Title')
    })

    test('should return result of subject.extractTechnical if provided and returns value', () => {
      type FormDataType = {
        subject: string
      }
      const extractTechnicalSubjectMock: SchemalessFormDataExtractor<FormDataType> = {
        type: 'schemaless',
        extractFn: (formData) => formData.subject,
      }

      const formDefinition = {
        title: 'Test Form Title',
        subject: {
          extractTechnical: extractTechnicalSubjectMock,
        },
      } as FormDefinitionSlovenskoSkGeneric
      const formData: FormDataType = {
        subject: 'Extracted Subject',
      }

      const result = extractFormSubjectTechnical(formDefinition, formData)
      expect(result).toBe('Extracted Subject')
    })

    test('should throw error when extraction returns empty result', () => {
      type FormDataType = {
        subject: string
      }
      const extractTechnicalSubjectMock: SchemalessFormDataExtractor<FormDataType> = {
        type: 'schemaless',
        extractFn: (formData) => formData.subject,
      }

      const formDefinition = {
        title: 'Test Form Title',
        subject: {
          extractTechnical: extractTechnicalSubjectMock,
        },
      } as FormDefinitionSlovenskoSkGeneric
      const formData = {
        wrongSubject: 'Extracted Subject',
      }

      expect(() => extractFormSubjectTechnical(formDefinition, formData)).toThrow()
    })

    getExampleFormPairs({
      formDefinitionFilterFn: isSlovenskoSkGenericFormDefinition,
    }).forEach(({ formDefinition, exampleForm }) => {
      test(`${exampleForm.name} GINIS subject should match snapshot`, () => {
        const result = extractFormSubjectTechnical(formDefinition, exampleForm.formData)
        expect(result).toMatchSnapshot()
      })
    })
  })

  describe('extractEmailFormEmail', () => {
    test('should return result of extractEmail', () => {
      const extractEmailMock: SchemalessFormDataExtractor<any> = {
        type: 'schemaless',
        extractFn: (formData) => formData.email,
      }

      const formDefinition = {
        type: FormDefinitionType.Email,
        email: {
          extractEmail: extractEmailMock,
        },
      } as FormDefinitionEmail
      const formData = {
        email: 'test@example.com',
      }

      const result = extractEmailFormEmail(formDefinition, formData)
      expect(result).toBe('test@example.com')
    })

    test('should throw error when extractEmailFormEmail on wrong data provided', () => {
      const extractEmailMock: SchemalessFormDataExtractor<any> = {
        type: 'schemaless',
        extractFn: (formData) => formData.email,
      }

      const formDefinition = {
        type: FormDefinitionType.Email,
        email: {
          extractEmail: extractEmailMock,
        },
      } as FormDefinitionEmail
      const formData = {
        wrongEmail: 'test@example.com',
      }

      expect(() => extractEmailFormEmail(formDefinition, formData)).toThrow()
    })

    getExampleFormPairs({
      formDefinitionFilterFn: isEmailFormDefinition,
    }).forEach(({ formDefinition, exampleForm }) => {
      test(`${exampleForm.name} email should match snapshot`, () => {
        const result = extractEmailFormEmail(formDefinition, exampleForm.formData)
        expect(result).toMatchSnapshot()
      })
    })
  })

  describe('extractEmailFormName', () => {
    test('should return undefined if extractName is not provided', () => {
      const formDefinition = {
        type: FormDefinitionType.Email,
        email: {},
      } as FormDefinitionEmail
      const formData = {}

      const result = extractEmailFormName(formDefinition, formData)
      expect(result).toBeUndefined()
    })

    test('should return result of extractName if provided', () => {
      const extractNameMock: SchemalessFormDataExtractor<any> = {
        type: 'schemaless',
        extractFn: (formData) => formData.name,
      }

      const formDefinition = {
        type: FormDefinitionType.Email,
        email: {
          extractName: extractNameMock,
        },
      } as FormDefinitionEmail
      const formData = {
        name: 'Test User',
      }

      const result = extractEmailFormName(formDefinition, formData)
      expect(result).toBe('Test User')
    })

    test('should throw error when extractEmailFormName on wrong data provided', () => {
      const extractNameMock: SchemalessFormDataExtractor<any> = {
        type: 'schemaless',
        extractFn: (formData) => formData.name,
      }

      const formDefinition = {
        type: FormDefinitionType.Email,
        email: {
          extractName: extractNameMock,
        },
      } as FormDefinitionEmail
      const formData = {
        wrongName: 'Test User',
      }

      expect(() => extractEmailFormName(formDefinition, formData)).toThrow()
    })

    getExampleFormPairs({
      formDefinitionFilterFn: isEmailFormDefinition,
    }).forEach(({ formDefinition, exampleForm }) => {
      test(`${exampleForm.name} name should match snapshot`, () => {
        const result = extractEmailFormName(formDefinition, exampleForm.formData)
        expect(result).toMatchSnapshot()
      })
    })
  })
})
