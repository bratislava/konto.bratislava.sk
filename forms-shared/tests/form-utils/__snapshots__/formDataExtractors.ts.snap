// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`formDataExtractors > extractEmailFormEmail > nahlaseniePodnetuKElektrickymKolobezkamExample email should match snapshot 1`] = `""`;

exports[`formDataExtractors > extractEmailFormEmail > oloDocistenieStanovistaZbernychNadobExample email should match snapshot 1`] = `"test@email.com"`;

exports[`formDataExtractors > extractEmailFormEmail > oloEnergetickeZhodnotenieOdpaduVZevoExample email should match snapshot 1`] = `"olo@test.sk"`;

exports[`formDataExtractors > extractEmailFormEmail > oloKoloTaxiExample email should match snapshot 1`] = `"test@email.com"`;

exports[`formDataExtractors > extractEmailFormEmail > oloMimoriadnyOdvozAZhodnotenieOdpaduExample email should match snapshot 1`] = `"test@email.com"`;

exports[`formDataExtractors > extractEmailFormEmail > oloOdvozObjemnehoOdpaduValnikomExample email should match snapshot 1`] = `"vah@odvoz.sk"`;

exports[`formDataExtractors > extractEmailFormEmail > oloOdvozOdpaduVelkokapacitnymAleboLisovacimKontajneromExample email should match snapshot 1`] = `"dolny@milan.com"`;

exports[`formDataExtractors > extractEmailFormEmail > oloOloTaxiExample email should match snapshot 1`] = `"test@email.com"`;

exports[`formDataExtractors > extractEmailFormEmail > oloPodnetyAPochvalyObcanovExample email should match snapshot 1`] = `"mrkva@jan.sk"`;

exports[`formDataExtractors > extractEmailFormEmail > oloTriedenyZberPapieraPlastovASklaPrePravnickeOsobyExample email should match snapshot 1`] = `"voda@odpad.com"`;

exports[`formDataExtractors > extractEmailFormEmail > oloTriedenyZberPapieraPlastovASklaPreSpravcovskeSpolocnostiExample email should match snapshot 1`] = `"test@test.test"`;

exports[`formDataExtractors > extractEmailFormEmail > oloUzatvorenieZmluvyONakladaniSOdpadom email should match snapshot 1`] = `"test@test.ts"`;

exports[`formDataExtractors > extractEmailFormName > nahlaseniePodnetuKElektrickymKolobezkamExample name should match snapshot 1`] = `undefined`;

exports[`formDataExtractors > extractEmailFormName > oloDocistenieStanovistaZbernychNadobExample name should match snapshot 1`] = `"Jozef"`;

exports[`formDataExtractors > extractEmailFormName > oloEnergetickeZhodnotenieOdpaduVZevoExample name should match snapshot 1`] = `"Jozef"`;

exports[`formDataExtractors > extractEmailFormName > oloKoloTaxiExample name should match snapshot 1`] = `"Odpad s.r.o."`;

exports[`formDataExtractors > extractEmailFormName > oloMimoriadnyOdvozAZhodnotenieOdpaduExample name should match snapshot 1`] = `"Jozef"`;

exports[`formDataExtractors > extractEmailFormName > oloOdvozObjemnehoOdpaduValnikomExample name should match snapshot 1`] = `"Odvoz odpadu a.s."`;

exports[`formDataExtractors > extractEmailFormName > oloOdvozOdpaduVelkokapacitnymAleboLisovacimKontajneromExample name should match snapshot 1`] = `"Správcovská a.s."`;

exports[`formDataExtractors > extractEmailFormName > oloOloTaxiExample name should match snapshot 1`] = `"Jozef"`;

exports[`formDataExtractors > extractEmailFormName > oloPodnetyAPochvalyObcanovExample name should match snapshot 1`] = `"Ján"`;

exports[`formDataExtractors > extractEmailFormName > oloTriedenyZberPapieraPlastovASklaPrePravnickeOsobyExample name should match snapshot 1`] = `"Odpad s.r.o."`;

exports[`formDataExtractors > extractEmailFormName > oloTriedenyZberPapieraPlastovASklaPreSpravcovskeSpolocnostiExample name should match snapshot 1`] = `"Test org"`;

exports[`formDataExtractors > extractEmailFormName > oloUzatvorenieZmluvyONakladaniSOdpadom name should match snapshot 1`] = `"Test"`;

exports[`formDataExtractors > extractFormSubjectPlain > komunitneZahradyExample form subject should match snapshot 1`] = `"Žiadosť o komunitnú záhradu - odporúčaný mestský pozemok"`;

exports[`formDataExtractors > extractFormSubjectPlain > nahlaseniePodnetuKElektrickymKolobezkamExample form subject should match snapshot 1`] = `"Nahlásenie podnetu k elektrickým kolobežkám"`;

exports[`formDataExtractors > extractFormSubjectPlain > oloDocistenieStanovistaZbernychNadobExample form subject should match snapshot 1`] = `"Dočistenie stanovišťa zberných nádob"`;

exports[`formDataExtractors > extractFormSubjectPlain > oloEnergetickeZhodnotenieOdpaduVZevoExample form subject should match snapshot 1`] = `"Energetické zhodnotenie odpadu v ZEVO"`;

exports[`formDataExtractors > extractFormSubjectPlain > oloKoloTaxiExample form subject should match snapshot 1`] = `"KOLO Taxi"`;

exports[`formDataExtractors > extractFormSubjectPlain > oloMimoriadnyOdvozAZhodnotenieOdpaduExample form subject should match snapshot 1`] = `"Mimoriadny odvoz a zhodnotenie odpadu"`;

exports[`formDataExtractors > extractFormSubjectPlain > oloOdvozObjemnehoOdpaduValnikomExample form subject should match snapshot 1`] = `"Odvoz objemného odpadu valníkom"`;

exports[`formDataExtractors > extractFormSubjectPlain > oloOdvozOdpaduVelkokapacitnymAleboLisovacimKontajneromExample form subject should match snapshot 1`] = `"Odvoz odpadu veľkokapacitným alebo lisovacím kontajnerom"`;

exports[`formDataExtractors > extractFormSubjectPlain > oloOloTaxiExample form subject should match snapshot 1`] = `"OLO Taxi"`;

exports[`formDataExtractors > extractFormSubjectPlain > oloPodnetyAPochvalyObcanovExample form subject should match snapshot 1`] = `"Podnety a pochvaly občanov"`;

exports[`formDataExtractors > extractFormSubjectPlain > oloTriedenyZberPapieraPlastovASklaPrePravnickeOsobyExample form subject should match snapshot 1`] = `"Triedený zber papiera, plastov a skla pre právnické osoby"`;

exports[`formDataExtractors > extractFormSubjectPlain > oloTriedenyZberPapieraPlastovASklaPreSpravcovskeSpolocnostiExample form subject should match snapshot 1`] = `"Triedený zber papiera, plastov a skla pre správcovské spoločnosti"`;

exports[`formDataExtractors > extractFormSubjectPlain > oloUzatvorenieZmluvyONakladaniSOdpadom form subject should match snapshot 1`] = `"Uzatvorenie zmluvy o nakladaní s odpadom"`;

exports[`formDataExtractors > extractFormSubjectPlain > oznamenieOPoplatkovejPovinnostiZaKomunalneOdpadyExample form subject should match snapshot 1`] = `"Oznámenie o poplatkovej povinnosti za komunálne odpady"`;

exports[`formDataExtractors > extractFormSubjectPlain > predzahradkyExample form subject should match snapshot 1`] = `"Registrácia novej predzáhradky"`;

exports[`formDataExtractors > extractFormSubjectPlain > priznanieKDaniZNehnutelnostiExample1 form subject should match snapshot 1`] = `"Priznanie k dani z nehnuteľností"`;

exports[`formDataExtractors > extractFormSubjectPlain > priznanieKDaniZNehnutelnostiExample2 form subject should match snapshot 1`] = `"Priznanie k dani z nehnuteľností"`;

exports[`formDataExtractors > extractFormSubjectPlain > priznanieKDaniZNehnutelnostiExample3 form subject should match snapshot 1`] = `"Priznanie k dani z nehnuteľností"`;

exports[`formDataExtractors > extractFormSubjectPlain > priznanieKDaniZNehnutelnostiExample4 form subject should match snapshot 1`] = `"Priznanie k dani z nehnuteľností"`;

exports[`formDataExtractors > extractFormSubjectPlain > priznanieKDaniZNehnutelnostiExample5 form subject should match snapshot 1`] = `"Priznanie k dani z nehnuteľností"`;

exports[`formDataExtractors > extractFormSubjectPlain > priznanieKDaniZNehnutelnostiExample5NoCalculators form subject should match snapshot 1`] = `"Priznanie k dani z nehnuteľností"`;

exports[`formDataExtractors > extractFormSubjectPlain > stanoviskoKInvesticnemuZameruExample form subject should match snapshot 1`] = `"Nový Bytový Dom"`;

exports[`formDataExtractors > extractFormSubjectPlain > zavazneStanoviskoKInvesticnejCinnostiExample form subject should match snapshot 1`] = `"Nový Bytový Dom"`;

exports[`formDataExtractors > extractFormSubjectPlain > ziadostONajomBytuExample form subject should match snapshot 1`] = `"Žiadosť o nájom bytu"`;

exports[`formDataExtractors > extractFormSubjectPlain > ziadostOSlobodnyPristupKInformaciamExample form subject should match snapshot 1`] = `"Žiadosť o sprístupnenie informácií podľa zákona č. 211/2000 Z.z."`;

exports[`formDataExtractors > extractFormSubjectPlain > ziadostOUzemnoplanovaciuInformaciuExample form subject should match snapshot 1`] = `"Žiadosť o územnoplánovaciu informáciu"`;

exports[`formDataExtractors > extractFormSubjectTechnical > komunitneZahradyExample GINIS subject should match snapshot 1`] = `"Žiadosť o komunitnú záhradu - odporúčaný mestský pozemok"`;

exports[`formDataExtractors > extractFormSubjectTechnical > oznamenieOPoplatkovejPovinnostiZaKomunalneOdpadyExample GINIS subject should match snapshot 1`] = `"Oznámenie o poplatkovej povinnosti za komunálne odpady"`;

exports[`formDataExtractors > extractFormSubjectTechnical > predzahradkyExample GINIS subject should match snapshot 1`] = `"Registrácia novej predzáhradky"`;

exports[`formDataExtractors > extractFormSubjectTechnical > stanoviskoKInvesticnemuZameruExample GINIS subject should match snapshot 1`] = `"e-SIZ Dunajská Nový Bytový Dom, p.č. 56789 kú Karlova Ves, Dúbravka"`;

exports[`formDataExtractors > extractFormSubjectTechnical > zavazneStanoviskoKInvesticnejCinnostiExample GINIS subject should match snapshot 1`] = `"e-ZST Dunajská Nový Bytový Dom, p.č. 56789 kú Karlova Ves, Dúbravka"`;

exports[`formDataExtractors > extractFormSubjectTechnical > ziadostONajomBytuExample GINIS subject should match snapshot 1`] = `"Žiadosť o nájom bytu"`;

exports[`formDataExtractors > extractFormSubjectTechnical > ziadostOSlobodnyPristupKInformaciamExample GINIS subject should match snapshot 1`] = `"Žiadosť o sprístupnenie informácií podľa zákona č. 211/2000 Z.z."`;

exports[`formDataExtractors > extractFormSubjectTechnical > ziadostOUzemnoplanovaciuInformaciuExample GINIS subject should match snapshot 1`] = `"Žiadosť o územnoplánovaciu informáciu"`;
